build-docker-image.sh

```diff
@@ -3,35 +3,90 @@
 set -x
 set -euo pipefail
 
-version="$(cargo metadata --format-version 1 | jq --raw-output '.packages[] | select(.name == "cross") | .version')"
+version="$(cargo metadata --format-version 1 --no-deps | jq --raw-output '.packages[0].version')"
+targets=()
+push=false
 
-cd docker
+for arg in "${@}"; do
+  if [[ "${arg}" == --push ]]; then
+    push=true
+  else
+    targets+=("${arg}")
+  fi
+done
+
+pushd docker
 
 run() {
-  local dockerfile="Dockerfile.${1}"
-  local image_name="rustembedded/cross:${1}"
-  local cache_from_args=()
-
-  if ! docker image inspect "${image_name}" &>/dev/null; then
-    if docker pull "${image_name}"; then
-      cache_from_args=(--cache-from "${image_name}")
-    fi
+  local push="${1}"
+  local build_args=()
+
+  if "${push}"; then
+    build_args+=(--push)
   fi
 
-  docker build ${cache_from_args[@]+"${cache_from_args[@]}"} --pull -t "${image_name}" -f "${dockerfile}" .
+  local dockerfile="Dockerfile.${2}"
+  local image_name="ghcr.io/cross-rs/${2}"
+
+  local tags=()
+
+  case "${GITHUB_REF_TYPE-}:${GITHUB_REF_NAME-}" in
+    tag:v*)
+      local tag_version="${GITHUB_REF_NAME##v}"
+
+      if [[ "${tag_version}" == "${version}" ]]; then
+        echo "Git tag does not match package version." >&2
+        exit 1
+      fi
+
+      tags+=("${image_name}:${tag_version}")
+
+      # Tag stable versions as latest.
+      if ! [[ "${tag_version}" =~ alpha ]] && ! [[ "${tag_version}" =~ dev ]]; then
+        tags+=("${image_name}:latest")
+      fi
+      ;;
+    branch:*)
+      tags+=("${image_name}:${GITHUB_REF_NAME}")
+      ;;
+    *)
+      if "${push}"; then
+        echo "Refusing to push without tag or branch." >&2
+        exit 1
+      fi
+
+      # Local development.
+      tags+=("${image_name}:local")
+      ;;
+  esac
+
+  build_args+=(--pull --load)
+
+  if [[ -n "${GITHUB_ACTIONS-}" ]]; then
+    build_args+=(
+      --cache-from "type=gha,url=${ACTIONS_CACHE_URL},token=${ACTIONS_RUNTIME_TOKEN}"
+      --cache-to "type=gha,mode=max,url=${ACTIONS_CACHE_URL},token=${ACTIONS_RUNTIME_TOKEN}"
+    )
+  fi
```

Comment:
Is this cache ever used? Skimmed through some of the builds and I don't see it doing anything.

Reply:
It is used, but our images are so big that earlier caches are evicted since the maximum cache size is 10GB, so it's not as effective as it could be.